shared/src/main/scala/mlscript/helpers.scala

```diff
@@ -16,6 +16,16 @@ abstract class TypeImpl extends Located { self: Type =>
     case Recursive(n, b) => n :: b.typeVarsList
     case _ => children.flatMap(_.typeVarsList)
   }
+
+  /**
+    * @return
+    *  set of non-recursive type variables in type
+    */
+  lazy val nonRecTypeVarSet: Set[TypeVar] = this match {
+    case Recursive(uv, body) => body.nonRecTypeVarSet - uv
+    case t@TypeVar(_, _) => Set(t)
```

Comment:
More efficient (avoids allocating a vararg):
```suggestion
    case t: TypeVar => Set.single(t)
```